File: leetcode/2140.解决智力问题.py
class Solution:
    def mostPoints(self, questions: List[List[int]]) -> int:
        n = len(questions)
        dp = [0] * (n + 1)   # 解决每道题及以后题目的最高分数
        for i in range(n - 1, -1, -1):
            dp[i] = max(dp[i + 1], questions[i][0] + dp[min(n, i + questions[i][1] + 1)])
        return dp[0]

File: leetcode/2140.解决智力问题.go
func mostPoints(questions [][]int) int64 {
    n := len(questions)
    dp := make([]int, n + 1)
    for i := n - 1; i >= 0; i-- {
        dp[i] = max(dp[i + 1], questions[i][0] + dp[min(n, i + questions[i][1] + 1)])
    }
    return int64(dp[0])
}